tests/test_black.py

```diff
@@ -1162,6 +1162,49 @@ def test_assert_equivalent_different_asts(self) -> None:
         with self.assertRaises(AssertionError):
             black.assert_equivalent("{}", "None")
 
+    def test_symlink_out_of_root_directory(self) -> None:
+        # prepare argumens
+        path = MagicMock()
+        root = THIS_DIR
+        child = MagicMock()
+        include = re.compile(black.DEFAULT_INCLUDES)
+        exclude = re.compile(black.DEFAULT_EXCLUDES)
+        report = black.Report()
+
+        # set the behavior of mock arguments
+        # child should behave like a symlink which resolved path is clearly
+        # outside of the root directory
+        path.iterdir.return_value = [child]
+        child.resolve.return_value = Path("/a/b/c")
+        child.is_symlink.return_value = True
+
+        # call the method
+        # it should not raise any error
+        list(black.gen_python_files_in_dir(path, root, include, exclude, report))
+
+        # check the call of the methods of the mock objects
+        path.iterdir.assert_called_once()
+        child.resolve.assert_called_once()
+        child.is_symlink.assert_called_once()
+
+        # set the behavior of mock arguments
+        # child should behave like a strange file which resolved path is clearly
+        # outside of the root directory
+        child.is_symlink.return_value = False
+
+        # call the method
+        # it should raise a ValueError
+        with self.assertRaises(ValueError):
+            list(black.gen_python_files_in_dir(path, root, include, exclude, report))
+
+        # check the call of the methods of the mock objects
+        path.iterdir.assert_called()
```

Comment:
Personally, I don't like test assertions like these, where the test deeply knows the tested method implementation.

I would suggest having simple assertions as if an error is raised or not based solely on the symlink validity. In other words, assert the behavior, not the implementation. That way, if the code changes in the future, but yields the same result, the tests would still be valid.

Reply:
I am using mock objects for the directory and the symlink inside it. I have set up a scenario to use them, according to their actual use within the function (which allows to perform the behavior test), so at the end of the test, I just check if the scenario was accurate.

I am not very experienced with mocking, but it seems to me that if I use mock objects, I somewhat make the test aware about the implementation of the tested function.

Reply:
You are correct on that point, just by using mocks you are already surfacing the inwards of the implementation, and thus asserting on the mock calls isn't wrong as well.

At work I have the same scenario on several (hundreds) tests, mocking 3rd-party calls (which relates the `os` file module here) at `setUp` for all test cases and only adding `side_effect` or new `return_value` when needed. And sometimes we assert on the mocks as you did. But we strive to assert on the behavior.

I guess my point here is mere semantics, and a little bit of food for thought. 😁 

Anyway, thank you for this PR, I was trying to use Black for the first time when I got this error, but it felt great to see someone already fixing it. 👍